src/main/java/org/opensearch/knn/index/query/ExactSearcher.java

```diff
@@ -71,15 +104,19 @@ private Map<Integer, Float> scoreAllDocs(KNNIterator iterator) throws IOExceptio
         return docToScore;
     }
 
-    private Map<Integer, Float> searchTopK(KNNIterator iterator, int k) throws IOException {
+    private Map<Integer, Float> searchTopCandidates(KNNIterator iterator, int limit, Predicate<Float> filterScore) throws IOException {
         // Creating min heap and init with MAX DocID and Score as -INF.
-        final HitQueue queue = new HitQueue(k, true);
+        final HitQueue queue = new HitQueue(limit, true);
         ScoreDoc topDoc = queue.top();
         final Map<Integer, Float> docToScore = new HashMap<>();
         int docId;
         while ((docId = iterator.nextDoc()) != DocIdSetIterator.NO_MORE_DOCS) {
-            if (iterator.score() > topDoc.score) {
-                topDoc.score = iterator.score();
+            final float currentScore = iterator.score();
+            if (filterScore != null && Predicate.not(filterScore).test(currentScore)) {
+                continue;
+            }
+            if (currentScore > topDoc.score) {
```

Comment:
compiler will make this `if () else { if() }` so this is not reducing nesting. would recommend making filter score @Nonnull and do `if (predicate.test && currentScore > topDoc.score)` if the intent is to reduce nesting

Reply:
updated